api/model/src/main/java/org/projectnessie/model/Util.java

```diff
@@ -25,45 +25,324 @@
+        case ZERO_BYTE:
+          e.append(DOT);
+          break;
+        default:
+          e.append(c);
+          break;
+      }
+    }
+    if (e.length() > 0) {
+      elements.add(e.toString());
+    }
+    return elements;
   }
 
   /**
-   * Convert these elements to a URL encoded path string.
+   * Convert these elements to a URI path/query string, which <em>may violate</em> <a
+   * href="https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#uri-path-canonicalization">Jakarta
+   * Servlet Spec 6, chapter 3.5.2 URI Path Canonicalization</a>, because it uses so-called control
+   * characters and ambiguous path elements.
    *
-   * @return String encoded for path use.
+   * <p>Users must prefer {@link #toPathStringEscaped(List)}, if the service support Nessie spec
+   * version 2.2.0 or newer, except for content-key related values in CEL filters.
+   *
+   * <p>Elements are separated by {@code .} characters. {@code .} characters in elements are
+   * replaced with the ASCII group separator (ASCII 29, {@code %29}).
+   *
+   * @param elements The content-key or namespace elements to encode.
+   * @return The URI path compatible representation of the given elements, possibly escaped. The
+   *     returned value should be URL-encoded before added to a URI path or query. The returned
+   *     value can be parsed with {@link #fromPathString(String)}.
    */
   public static String toPathString(List<String> elements) {
-    return elements.stream()
-        .map(x -> x.replace(DOT, GROUP_SEPARATOR).replace(ZERO_BYTE, GROUP_SEPARATOR))
-        .collect(Collectors.joining("."));
+    StringBuilder sb = new StringBuilder();
+    for (String element : elements) {
+      if (sb.length() > 0) {
+        sb.append('.');
+      }
+      int l = element.length();
+      for (int i = 0; i < l; i++) {
+        char c = element.charAt(i);
+        sb.append(c == DOT || c == ZERO_BYTE ? GROUP_SEPARATOR : c);
+      }
+    }
+    return sb.toString();
+  }
+
+  /**
+   * Escapes content-key elements into a URI path/query compatible form that does not violate <a
+   * href="https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#uri-path-canonicalization">Jakarta
+   * Servlet Spec 6, chapter 3.5.2 URI Path Canonicalization</a> by escaping the characters {@code
+   * /}, {@code \}, {@code %}, and if escaping also the {@code .}.
+   *
+   * <p>Algorithm:
+   *
+   * <ul>
+   *   <li>Elements are separated using a single {@code .} character.
+   *   <li>If an element starts with a {@code .} character, that element <b>and all following
+   *       elements</b> use the "problematic character escaping":
+   *       <ul>
+   *         <li>a {@code .} is escaped as {@code ._}
+   *         <li>a {@code /} is escaped as <code>.{</code>
+   *         <li>a {@code \} is escaped as <code>.}</code>
+   *         <li>a {@code %} is escaped as {@code .[}
+   *       </ul>
+   * </ul>
+   *
+   * <p>Some examples:
```

Comment:
I think I found a situation where this encoding scheme does not work:

```java
  @Test
  void roundTrip() {
    List<String> expected = List.of("abc.", "{def");
    String escaped = Util.toPathStringEscaped(expected);
    List<String> actual = Util.fromPathString(escaped);
    soft.assertThat(actual).isEqualTo(expected);
  }
```

```
org.opentest4j.AssertionFailedError: 
expected: ["abc.", "{def"]
 but was: ["abc./def"]
```

Reply:
Right - seems like it's necessary to use `..` when it's escaping.